### PR TITLE
🚧 Map Concepts between KF and FHIR

### PR DESCRIPTION
**Pull request by liberaliscomputing**
This PR is part of an ongoing effort to map Kids First concepts and FHIR resources/profiles. The relevant ticket is #74.

### CODE REVIEW

File: docs/no_peeking/mappings/specimen.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+
+specimen = {
+    'resourceType': 'Specimen',
+    'id': BIOSPECIMEN.TARGET_SERVICE_ID,
+    'biosample-individual-age-at-collection': None,
+    'biosampe-histological-diagnosis': None,
+    'biosample-tumor-progression': None,
+    'biosample-tumor-grade': None,
+    'biosample-control': True or False, # defaults to True
```
Comment (fiendish) [inline docs/no_peeking/mappings/specimen.py]: can we get comments on some of these ambiguous ones?

File: docs/no_peeking/mappings/patient.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        return 'other'
+    elif x == constants.COMMON.UNKNOWN:
+        return 'unknown'
+    else:
+        raise Exception('Unknown Patient gender')
+
+
+patient = {
+    'resourceType': 'Patient'
+    'id': PARTICIPANT.TARGET_SERVICE_ID,
```
Comment (fiendish) [inline docs/no_peeking/mappings/patient.py]: target service ID isn't known before it goes into the server. what do we do about this?

Comment (liberaliscomputing) [reply docs/no_peeking/mappings/patient.py]: Yeah, the use of `TARGET_SERVICE_ID` is like place-holding at the moment. As far as I understand, the `'id'` field is the one used when we hit each of the REST endpoints like `BASE_URL/Patient/id`. So, it seems like we need to define some consistent, global scheme of `'id'`. Any thoughts on this, @znatty22 and @ShahimEssaid?

File: docs/no_peeking/mappings/specimen.py (hunk 1 above)
Comment (liberaliscomputing) [reply docs/no_peeking/mappings/specimen.py]: I will provide a more complete README.md describing my findings and rationale for this work as well as docstrings per every mapper.

Comment (liberaliscomputing) [reply docs/no_peeking/mappings/specimen.py]: And could you clarify what "these ambiguous ones" means, @fiendish?

Comment (fiendish) [reply docs/no_peeking/mappings/specimen.py]: so like it says "True or False" which is of course logically always True. what is the field actually supposed to have in it?

Comment (liberaliscomputing) [reply docs/no_peeking/mappings/specimen.py]: Aha, I see what you mean. I will also add some inline comments across all the mappers if the assignment of a certain value is not clear.

File: docs/no_peeking/mappings/patient.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
+def individual_texonomy(x):
+    """
+    https://aehrc.github.io/fhir-phenopackets-ig/ValueSet-taxonomy.html
+    """
+    if x == constants.SPECIES.DOG:
+        return {
+            'code': 'NCBITaxon_9615',
```
Comment (fiendish) [inline docs/no_peeking/mappings/patient.py]: should these be : instead of _?

Comment (liberaliscomputing) [reply docs/no_peeking/mappings/patient.py]: "_" is changed to ":".

File: docs/no_peeking/mappings/specimen.py (hunk 1 above)
Comment (liberaliscomputing) [reply docs/no_peeking/mappings/specimen.py]: The required changes have been made!

**Comment by fiendish** [comment]
```
 'extension': [
         {
             'text': 'CodedOnset',
             'url': 'http://ga4gh.org/fhir/phenopackets/StructureDefinition/CodedOnset',
             'valueCodeableConcept': {
                 'coding': [
                     {
                         'system': 'http://purl.obolibrary.org/obo/hp.owl',
                         'code': 'HP:0410280',
                         'display': 'Pediatric onset'
                     }
                 ]
             }
         }
     ],
```

doesn't look like 

```
'extension': [
        {
            'text': 'coded-onset',
            'coding': [
                {
                    'system': 'http://foo/coded-onset',
                    'code': 'bar'
                    'display': 'Pediatric onset'

                }
            ]
        },
```

@znatty22 ?

File: docs/no_peeking/mappings/biosample.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            'display': 'Tissue'
+        })
+    else:
+        raise Exception('Unknown Biosample type')
+
+
+biosample = {
+    'resourceType': 'Specimen',
+    'id': BIOSPECIMEN.TARGET_SERVICE_ID,
+    'meta: {
```
Comment (fiendish) [inline docs/no_peeking/mappings/biosample.py]: missing close quote

File: docs/no_peeking/mappings/disease.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    """
+    http://hl7.org/fhir/R4/valueset-body-site.html
+    """
+    pass
+
+
+disease =  {
+    'resourceType': 'Condition',
+    'id': DIAGNOSIS.TARGET_SERVICE_ID,
+    'meta: {
```
Comment (fiendish) [inline docs/no_peeking/mappings/disease.py]: missing close quote

File: docs/no_peeking/mappings/family.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    else:
+        raise Exception('Unknown Family family-member-type')
+
+def family_member_phenopacket(x): pass
+
+
+family = {
+    'resourceType': 'Group',
+    'id': FAMILY.TARGET_SERVICE_ID,
+    'meta: {
```
Comment (fiendish) [inline docs/no_peeking/mappings/family.py]: missing close quote

File: docs/no_peeking/mappings/individual.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    elif x == constants.COMMON.UNKNOWN:
+        return 'unknown'
+    else:
+        raise Exception('Unknown Individual gender')
+
+
+individual = {
+    'resourceType': 'Patient',
+    'id': PARTICIPANT.TARGET_SERVICE_ID,
+    'meta: {
```
Comment (fiendish) [inline docs/no_peeking/mappings/individual.py]: missing close quote

File: docs/no_peeking/mappings/pedigree_node.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+    elif x == constants.COMMON.UNKNOWN:
+        return 'unknown'
+    else:
+        raise Exception('Unknown FamilyMemberHistory sex')
+
+
+pedigree_node = {
+    'resourceType': 'FamilyMemberHistory',
+    'id': FAMILY_RELATIONSHIP.TARGET_SERVICE_ID,
+    'meta: {
```
Comment (fiendish) [inline docs/no_peeking/mappings/pedigree_node.py]: missing close quote

File: docs/no_peeking/mappings/phenotypic_feature.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+            'code': 'POS',
+            'display': 'Positive'
+        })
+    else:
+        raise Exception('Unknown PhenotypicFeature interpretation')
+
+
+phenotypic_feature = {
+    'resourceType': 'Observation',
+    'meta: {
```
Comment (fiendish) [inline docs/no_peeking/mappings/phenotypic_feature.py]: missing close quote

File: docs/no_peeking/mappings/biosample.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+            },
+            {
+                'value': PARTICIPANT.ID
+            }
+        ],
+        'display': PARTICIPANT.ID
+    },
+    'collection': {
+        'quantity': {
+            'value': BIOSPECIMEN.VOLUME_UL
```
Comment (fiendish) [inline docs/no_peeking/mappings/biosample.py]: missing comma

File: docs/no_peeking/mappings/biosample.py (hunk 4 above)
Comment (liberaliscomputing) [reply docs/no_peeking/mappings/biosample.py]: fixed!

File: docs/no_peeking/mappings/disease.py (hunk 5 above)
Comment (liberaliscomputing) [reply docs/no_peeking/mappings/disease.py]: fixed!

File: docs/no_peeking/mappings/family.py (hunk 6 above)
Comment (liberaliscomputing) [reply docs/no_peeking/mappings/family.py]: fixed!

File: docs/no_peeking/mappings/individual.py (hunk 7 above)
Comment (liberaliscomputing) [reply docs/no_peeking/mappings/individual.py]: fixed!

File: docs/no_peeking/mappings/pedigree_node.py (hunk 8 above)
Comment (liberaliscomputing) [reply docs/no_peeking/mappings/pedigree_node.py]: fixed!

File: docs/no_peeking/mappings/phenotypic_feature.py (hunk 9 above)
Comment (liberaliscomputing) [reply docs/no_peeking/mappings/phenotypic_feature.py]: fixed!

File: docs/no_peeking/mappings/biosample.py (hunk 10 above)
Comment (liberaliscomputing) [reply docs/no_peeking/mappings/biosample.py]: fixed!

**Comment by znatty22** [comment]
@fiendish @liberaliscomputing 

Oops, did that too hastily. The first one is correct.

File: docs/no_peeking/mappings/individual.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+
+individual = {
+    'resourceType': 'Patient',
+    'id': CONCEPT.PARTICIPANT.TARGET_SERVICE_ID,
+    'meta': {
+        'profile': ['http://ga4gh.fhir.phenopackets/StructureDefinition/Individual']
+    },
+    'extension': [
+        {
+            'text': 'KaryotypicSex',
```
Comment (ShahimEssaid) [inline docs/no_peeking/mappings/individual.py]: What's the 'text' field for? I'm guessing for documentation. It doesn't have a place in Extension.

File: docs/no_peeking/mappings/individual.py [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+            }
+        }
+    ],
+    'identifier': [
+        {
+            'system': 'https://kf-api-dataservice.kidsfirstdrc.org/participants',
+            'value': CONCEPT.PARTICIPANT.TARGET_SERVICE_ID
+        },
+        {
+            'value': CONCEPT.PARTICIPANT.ID
```
Comment (ShahimEssaid) [inline docs/no_peeking/mappings/individual.py]: This could use a 'system' as well

File: docs/no_peeking/mappings/biosample.py [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+    if x == constants.COMMON.FALSE:
+        return 'unavailable'
+    elif x == constants.COMMON.TRUE:
+        return 'available'
+    else:
+        raise Exception('Unknown Biosample status')
+
+def biosample_type(x):
+    """
+    http://terminology.hl7.org/ValueSet/v2-0487
```
Comment (ShahimEssaid) [inline docs/no_peeking/mappings/biosample.py]: How are these URLs in the comments used?  This URL doesn't resolve but it is taken from http://hl7.org/fhir/R4/v2/0487/index.html as the CodeSystem URL.

File: docs/no_peeking/mappings/biosample.py [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+        }
+    else:
+        raise Exception('Unknown Biosample type')
+
+
+biosample = {
+    'resourceType': 'Specimen',
+    'id': CONCEPT.BIOSPECIMEN.TARGET_SERVICE_ID,
+    'meta': {
+        'profile': ['http://ga4gh.fhir.phenopackets/StructureDefinition/Biosample']
```
Comment (ShahimEssaid) [inline docs/no_peeking/mappings/biosample.py]: We could have placeholder KF profiles based on the PK profiles to give us more flexibility down the road. We can also copy their "differential" element definitions to our KF profiles (so we have similar and interoperable constraints) and keep our profiles based on the base FHIR profiles.

File: docs/no_peeking/mappings/biosample.py [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+            {
+                'value': CONCEPT.PARTICIPANT.ID
+            }
+        ],
+        'display': CONCEPT.PARTICIPANT.ID
+    },
+    'collection': {
+        'quantity': {
+            'value': CONCEPT.BIOSPECIMEN.VOLUME_UL,
+            'unit': 'uL'
```
Comment (ShahimEssaid) [inline docs/no_peeking/mappings/biosample.py]: We should also add a system for the unit. Not sure why the unit wasn't a coding/CodeableConcept in FHIR.

File: docs/no_peeking/mappings/disease.py [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+                'system': 'https://kf-api-dataservice.kidsfirstdrc.org/participants',
+                'value': CONCEPT.PARTICIPANT.TARGET_SERVICE_ID
+            },
+            {
+                'value': CONCEPT.PARTICIPANT.ID
+            }
+        ],
+        'display': CONCEPT.PARTICIPANT.ID
+    },
+    'onsetAge': CONCEPT.DIAGNOSIS.EVENT_AGE_DAYS
```
Comment (ShahimEssaid) [inline docs/no_peeking/mappings/disease.py]: What's the format of KF dates? This has to be coded as a FHIR Quantity if it's days.

http://hl7.org/fhir/R4/datatypes.html#quantity

Comment (fiendish) [reply docs/no_peeking/mappings/disease.py]: KF doesn't have dates ever. We only have ages.

File: docs/no_peeking/mappings/README.md [hunk 17]
```diff
@@ -0,0 +1,28 @@
+# Concept-Mappers between Kids First and FHIR
```
Comment (znatty22) [inline docs/no_peeking/mappings/README.md]:
```suggestion
# Concept-Mappers between Kids First and Phenopackets FHIR
```

File: docs/no_peeking/mappings/biosample.py [hunk 18]
```diff
@@ -0,0 +1,94 @@
+"""
+This module maps Kids First biospecimen to Phenopackets Biosample (derived from FHIR Specimen).
+Please visit https://aehrc.github.io/fhir-phenopackets-ig/StructureDefinition-Biosample.html
+    for the detailed structure definition.
```
Comment (znatty22) [inline docs/no_peeking/mappings/biosample.py]:
```suggestion
for the detailed structure definition.
```

File: docs/no_peeking/mappings/biosample.py [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+    ],
+    'status': biosample_status(CONCEPT.BIOSPECIMEN.VISIBLE),
+    'type': {
+        'coding': [
+            biosample_type(CONCEPT.BIOSPECIMEN.COMPOSITION)
+        ],
+        'text': CONCEPT.BIOSPECIMEN.COMPOSITION
+    },
+    'subject': {
+        'reference': f'Individual/CONCEPT.{PARTICIPANT.TARGET_SERVICE_ID}',
```
Comment (znatty22) [inline docs/no_peeking/mappings/biosample.py]: Should this be:
```suggestion
        'reference': f'Patient/CONCEPT.{PARTICIPANT.TARGET_SERVICE_ID}',
```
I think the references have to have the FHIR base type in the prefix not the id of the StructureDefinition...I might be wrong. Could someone check the spec and verify?

File: docs/no_peeking/mappings/biosample.py [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+        'type': 'Individual',
```
Comment (znatty22) [inline docs/no_peeking/mappings/biosample.py]: Should this be (same reason as above)
```suggestion
        'type': 'Patient',
```

File: docs/no_peeking/mappings/individual.py (hunk 11 above)
Comment (znatty22) [reply docs/no_peeking/mappings/individual.py]: Agreed, there is no `text` field in extension but there is one in the CodeableConcept. Move the `text` field in the CodeableConcept like this `valueCodeableConcept.text`

File: docs/no_peeking/mappings/individual.py [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+            'url': 'http://ga4gh.org/fhir/phenopackets/StructureDefinition/KaryotypicSex',
+            'valueCodeableConcept': {
+                'coding': [
+                    karyotypic_sex(CONCEPT.PARTICIPANT.GENDER)
+                ],
+                'text': CONCEPT.PARTICIPANT.GENDER
+            }
+        },
+        {
+            'text': 'Taxonomy',
```
Comment (znatty22) [inline docs/no_peeking/mappings/individual.py]: Same comment as above

File: docs/no_peeking/mappings/individual.py (hunk 11 above)
Comment (liberaliscomputing) [reply docs/no_peeking/mappings/individual.py]: @fiendish will create a PR to this branch which fixes this issue!

File: docs/no_peeking/mappings/individual.py (hunk 12 above)
Comment (liberaliscomputing) [reply docs/no_peeking/mappings/individual.py]: @fiendish will create a PR to this branch which adds a system attribute to the second set of IDs.

File: docs/no_peeking/mappings/biosample.py (hunk 13 above)
Comment (liberaliscomputing) [reply docs/no_peeking/mappings/biosample.py]: For all the custom functions defined, the URLs in the docstrings are not for direct use, but for reference. And, when I created this PR, it was before the Phenopackets team has made the recent IG updates and the URLs were valid. 

@fiendish will create a PR to this branch which resolves terminology issues.

File: docs/no_peeking/mappings/family.py [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+    'extension': [
+        {
+            'text': 'PedigreeNodeReference',
+            'url': 'http://ga4gh.org/fhir/phenopackets/StructureDefinition/PedigreeNodeReference',
+            'valueReference': None
+        }
+    ],
+    'type': family_type(CONCEPT.PARTICIPANT.SPECIES or constants.SPECIES.HUMAN),
+    'actual': constants.COMMON.TRUE, # defaults to True
+    'member': [
```
Comment (ShahimEssaid) [inline docs/no_peeking/mappings/family.py]: The "member" is missing the "family-member-phenopacket" extension shown here:

https://aehrc.github.io/fhir-phenopackets-ig/StructureDefinition-Family-definitions.html#Group.member.extension:family-member-phenopacket

Also, the new IG doesn't show the ValueSet/CodeSystme  that is used in the "family-member-type" extension. I think it was shown in the older IG format.

File: docs/no_peeking/mappings/pedigree_node.py [hunk 23]
```diff
@@ -0,0 +1,153 @@
...
+
+pedigree_node = {
+    'resourceType': 'FamilyMemberHistory',
+    'id': CONCEPT.FAMILY_RELATIONSHIP.TARGET_SERVICE_ID,
+    'meta': {
+        'profile': ['http://ga4gh.org/fhir/phenopackets/StructureDefinition/PedigreeNode']
+    },
+    'extension': [
+        {
+            'text': 'AffectedStatus',
```
Comment (ShahimEssaid) [inline docs/no_peeking/mappings/pedigree_node.py]: Same comment as in other files regarding "text" not being directly under Extension

File: docs/no_peeking/mappings/pedigree_node.py [hunk 24]
```diff
@@ -0,0 +1,153 @@
...
+            'url': 'http://ga4gh.org/fhir/phenopackets/StructureDefinition/AffectedStatus',
+            'valueCoding': [
+                affected_status(CONCEPT.PARTICIPANT.IS_AFFECTED_UNDER_STUDY)
+            ]
+        },
+        {
+            'text': 'FamilyIdentifier',
```
Comment (ShahimEssaid) [inline docs/no_peeking/mappings/pedigree_node.py]: Same

File: docs/no_peeking/mappings/pedigree_node.py [hunk 25]
```diff
@@ -0,0 +1,153 @@
...
+                    'system': 'https://kf-api-dataservice.kidsfirstdrc.org/families',
+                    'value': CONCEPT.FAMILY.TARGET_SERVICE_ID
+                },
+                {
+                    'value': CONCEPT.FAMILY.ID
+                }
+            ]
+        },
+        {
+            'text': 'IndividualReference',
```
Comment (ShahimEssaid) [inline docs/no_peeking/mappings/pedigree_node.py]: Same

File: docs/no_peeking/mappings/pedigree_node.py [hunk 26]
```diff
@@ -0,0 +1,153 @@
...
+            'valueReference': {
+                'reference': f'Individual/{CONCEPT.FAMILY_RELATIONSHIP.PERSON1.TARGET_SERVICE_ID}',
+                'type': 'Individual',
+                'identifier': [
+                    {
+                        'system': 'https://kf-api-dataservice.kidsfirstdrc.org/participants',
+                        'value': CONCEPT.FAMILY_RELATIONSHIP.PERSON1.TARGET_SERVICE_ID
+                    },
+                    {
+                        'value': CONCEPT.FAMILY_RELATIONSHIP.PERSON1.ID
```
Comment (ShahimEssaid) [inline docs/no_peeking/mappings/pedigree_node.py]: Add "system"

File: docs/no_peeking/mappings/pedigree_node.py [hunk 27]
```diff
@@ -0,0 +1,153 @@
...
+                }
+            ]
+        },
+        {
+            'text': 'IndividualReference',
+            'url': 'http://ga4gh.org/fhir/phenopackets/StructureDefinition/IndividualReference',
+            'valueReference': {
+                'reference': f'Individual/{CONCEPT.FAMILY_RELATIONSHIP.PERSON1.TARGET_SERVICE_ID}',
+                'type': 'Individual',
+                'identifier': [
```
Comment (ShahimEssaid) [inline docs/no_peeking/mappings/pedigree_node.py]: This isn't really necessary if there is a direct reference. See: http://hl7.org/fhir/R4/references-definitions.html#Reference.identifier

File: docs/no_peeking/mappings/pedigree_node.py [hunk 28]
```diff
@@ -0,0 +1,153 @@
...
+            }
+        }
+    ],
+    'identifier': [
+        {
+            'system': 'https://kf-api-dataservice.kidsfirstdrc.org/family-relationships',
+            'value': CONCEPT.FAMILY_RELATIONSHIP.TARGET_SERVICE_ID
+        },
+        {
+            'value': CONCEPT.FAMILY_RELATIONSHIP.ID
```
Comment (ShahimEssaid) [inline docs/no_peeking/mappings/pedigree_node.py]: Add 'system'

File: docs/no_peeking/mappings/pedigree_node.py [hunk 29]
```diff
@@ -0,0 +1,153 @@
...
+        }
+    ],
+    'status': None, # pedigree_node_status()
+    'patient': {
+        'reference': f'Individual/{CONCEPT.FAMILY_RELATIONSHIP.PERSON2.TARGET_SERVICE_ID}',
+        'type': 'Individual',
+        'identifier': [
```
Comment (ShahimEssaid) [inline docs/no_peeking/mappings/pedigree_node.py]: same as above. not needed if 'reference' is available.

File: docs/no_peeking/mappings/biosample.py [hunk 30]
```diff
@@ -63,24 +67,36 @@ def yield_biosamples(eng, table, study_id, individuals):
...
+                {
+                    "system": "http://kf-api-dataservice.kidsfirstdrc.org/biospecimens", "value": kfid
+                }
+            )
+
         if subject_id:
             retval["subject"] = {
-                "reference": f"Patient/{individuals[subject_id]['id']}"
+                "reference": f"Patient/{individuals[subject_id]['id']}",
+                "type": iRType
```
Comment (fiendish) [inline docs/no_peeking/mappings/biosample.py]: what is this type field for?

File: .gitignore [hunk 31]
```diff
@@ -115,3 +115,4 @@ logsettings.env
 
 # Miscellaneous
 .vscode/
+.DS_Store
```
Comment (fiendish) [inline .gitignore]: Can everything starting with . in this file just be replaced by `.*` instead?

**Pull request closed without merging**